arch: arc: update arc connect debug select cpu mask when new cpu come online

IRISZZW opened the pull request:
The ARConnect Inter-core Debug Unit (ICD) provides additional
debug assist features in multi-core scenarios. It's useful to halt
other cores when one core is halted.

Before we program ICD in master core(core 0) initial stage, add
all cores to mask. so we need to make sure other slave cores have
launched and in running mode before we enable ICD in master core.

If we launch master core first, then launch slave cores by master core
conditionally, in this scenario, it's not OK.

Let's update arc connect debug (ARConnect ICD) select mask when
new slave core come online by slave core self, instead of use hardcoded
select mask.

ruuddw commented:
Great, this looks like a much more robust solution than #36535.
Can you please add a description of the problem this PR solves (like in the other PR), or even better create an issue for that?

abrodkin reviewed (changes requested):
While implementation LGTM I'd like to see a rationale for this change clearly stated in the commit message. I.e. why do we want this way and why it's better than what we used to have.

evgeniy-paltsev commented:
Should we revert #36535 (it has already been merged)?

evgeniy-paltsev commented on arch/arc/core/arc_smp.c:
```diff
@@ -66,17 +66,40 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	z_arc_connect_ici_clear();
 	z_irq_priority_set(IRQ_ICI, ARCV2_ICI_IRQ_PRIORITY, 0);
 	irq_enable(IRQ_ICI);
+
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
+		arc_connect_debug_setup(cpu_num);
+	}
```
I'd place this right after we read BCR, so we update the debug mask as early as possible.

evgeniy-paltsev commented on arch/arc/core/arc_smp.c:
```diff
@@ -66,17 +66,40 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 }
 
+#ifdef CONFIG_SMP
+static void arc_connect_debug_setup(int cpu_num)
```
Just a nitpick - probably name it something like `arc_connect_debug_mask_update` so something similar as we call it multiple times.

ruuddw commented:
> Should we revert #36535 (it has already been merged)?

I thought about adding a DNM, but didn't do that since the PRs are orthogonal and reversing the order doesn't hurt.

abrodkin commented:
@evgeniy-paltsev agree with @ruuddw - these are orthogonal changes and start of the master in the very end makes sense IMHO anyways.

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -66,17 +66,41 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+	bcr.val = z_arc_v2_aux_reg_read(_ARC_V2_CONNECT_BCR);
+
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
```
Either move this comment above `if` or add another identation level, i.e. move right to align with the next line `arc_connect_debug_mask_update(cpu_num);`.

IRISZZW replied:
Fixed.

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -66,17 +66,41 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	z_arc_connect_ici_clear();
 	z_irq_priority_set(IRQ_ICI, ARCV2_ICI_IRQ_PRIORITY, 0);
 	irq_enable(IRQ_ICI);
+
```
Why adding this newline?

IRISZZW replied:
Removed.

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -117,6 +141,11 @@ static int arc_smp_init(const struct device *dev)
 	bcr.val = z_arc_v2_aux_reg_read(_ARC_V2_CONNECT_BCR);
 
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
```
Ditto

IRISZZW replied:
Ditto.

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -66,12 +66,35 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+
+	core_mask |= z_arc_connect_debug_select_read();
+	z_arc_connect_debug_select_set(core_mask);
+	/* Debugger halt cores at conditions */
```
1. Debugger halt**s**
2. Could you please list conditions while we're at it?

IRISZZW committed:
arch: arc: update ARConnect ICD select mask when new cpu come online

The ARConnect Inter-core Debug Unit (ICD) provides additional
debug assist features in multi-core scenarios. It's useful to halt
other cores when one core is halted.

Before we program ICD in master core(core 0) initial stage, add
all cores to mask. so we need to make sure other slave cores have
launched and in running mode before we enable ICD in master core.

If we launch master core first, then launch slave cores by master
core conditionally, in this scenario, it's not OK.

Let's update arc connect debug (ARConnect ICD) select mask
when new slave core come online by slave core self, instead of
use hardcoded select mask.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
